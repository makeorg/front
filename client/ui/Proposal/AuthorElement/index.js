import * as React from 'react';
import { i18n } from 'Shared/i18n';
import { type Author } from 'Shared/types/proposal';
import { DateHelper } from 'Shared/helpers/date';
import { getOrganisationProfileLink } from 'Shared/helpers/url';
import { Avatar } from 'Client/ui/Avatar';
import { SvgCheckedSymbol } from 'Client/ui/Svg/elements';
import { RedLinkStyle } from 'Client/ui/Elements/LinkElements';
import { TextColors } from 'Client/app/assets/vars/Colors';
import { ScreenReaderItemStyle } from 'Client/ui/Elements/AccessibilityElements';
import {
  AuthorDescriptionStyle,
  AuthorInfosStyle,
  AuthorSeparatorStyle,
  ProposalStatusStyle,
} from './Styled';

type Props = {
  /** Object with author's properties */
  author: Author,
  /** Country of the proposal */
  country: string,
  /** Language of the proposal */
  language: string,
  /** Date of creation of proposal */
  createdAt?: string,
  /** Include avatar */
  withAvatar: boolean,
};

const ProposalAuthorAge = ({ age }) => {
  if (!age) {
    return null;
  }

  return (
    <React.Fragment>
      {`, ${i18n.t('proposal_card.author.age', { age })}`}
    </React.Fragment>
  );
};

export const ProposalAuthorElement = (props: Props) => {
  const {
    author,
    country,
    language,
    createdAt,
    withAvatar,
    withStatus,
    formattedProposalStatus,
  } = props;

  return (
    <AuthorDescriptionStyle>
      <AuthorInfosStyle withAvatar={withAvatar}>
        {withAvatar && (
          <React.Fragment>
            <Avatar>
              {author.avatarUrl && <img src={author.avatarUrl} alt="" />}
            </Avatar>
            &nbsp;
          </React.Fragment>
        )}
        <ScreenReaderItemStyle>
          {i18n.t('proposal_card.author.from')}
        </ScreenReaderItemStyle>
        {author.organisationName ? (
          <React.Fragment>
            <RedLinkStyle
              href={getOrganisationProfileLink(
                country,
                language,
                author.organisationSlug
              )}
            >
              {author.organisationName}
            </RedLinkStyle>
            <SvgCheckedSymbol
              style={{
                fontSize: '14px',
                marginLeft: '5px',
                fill: TextColors.Blue,
              }}
            />
          </React.Fragment>
        ) : (
          author.firstName
        )}
        <ProposalAuthorAge age={author.age} />
        {!!createdAt && (
          <React.Fragment>
            <AuthorSeparatorStyle>&nbsp;&bull;&nbsp;</AuthorSeparatorStyle>
            <ScreenReaderItemStyle>
              {i18n.t('proposal_card.author.date')}
            </ScreenReaderItemStyle>
            <time dateTime={createdAt}>
              {DateHelper.creationDateFormat(createdAt)}
            </time>
          </React.Fragment>
        )}
      </AuthorInfosStyle>
      {withStatus && (
        <ProposalStatusStyle className={`status-${formattedProposalStatus}`}>
          <ScreenReaderItemStyle>
            {i18n.t('proposal_card.status.title')}
          </ScreenReaderItemStyle>
          {i18n.t(`proposal_card.status.${formattedProposalStatus}`)}
        </ProposalStatusStyle>
      )}
    </AuthorDescriptionStyle>
  );
};

ProposalAuthorElement.defaultProps = {
  withAvatar: false,
  withStatus: false,
  formattedProposalStatus: undefined,
};
