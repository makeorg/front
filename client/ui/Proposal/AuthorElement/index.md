##### ProposalAuthorElement Without Avatar Example :
```jsx
import { proposalTypeFixture } from '../../../../shared/types/__fixtures__/proposal.fixture.js';
import { ProposalAuthorElement } from './index';
  <ProposalAuthorElement
    author={proposalTypeFixture.author}
    country={proposalTypeFixture.country}
    language={proposalTypeFixture.language}
    createdAt={proposalTypeFixture.createdAt}
  />
```

##### ProposalAuthorElement With Avatar Example :
```jsx
import { proposalTypeFixture } from '../../../../shared/types/__fixtures__/proposal.fixture.js';
import { ProposalAuthorElement } from './index';
  <ProposalAuthorElement
    author={proposalTypeFixture.author}
    country={proposalTypeFixture.country}
    language={proposalTypeFixture.language}
    createdAt={proposalTypeFixture.createdAt}
    withAvatar
    formattedProposalStatus={proposalTypeFixture.status.lowercase}
  />
```