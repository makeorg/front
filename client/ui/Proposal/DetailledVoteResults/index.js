// @flow
import React, { useRef } from 'react';
import { getTotalVotesCount, getVotesPercent } from 'Shared/helpers/voteResult';
import { type Vote as TypeVote } from 'Shared/types/proposal';
import { ScreenReaderItemStyle } from 'Client/ui/Elements/AccessibilityElements';
import { i18n } from 'Shared/i18n';
import { useMobile } from 'Client/hooks/useMedia';
import { type TypeSliderParams } from 'Shared/types/views';
import { useSlider } from 'Client/hooks/useSlider';
import { GliderStylesheet } from 'Client/app/assets/css-in-js/GliderStyle';
import { DetailledResultItem } from './Item';
import { VoteProgress } from './Progress';
import { DetailledItemListStyle } from './style';

type Props = {
  votes: TypeVote[],
  proposalId: string,
};

export const DetailledVoteResults = (props: Props) => {
  const { votes, proposalId } = props;
  const totalVotesCount = getTotalVotesCount(votes);
  const votesPercent = getVotesPercent(votes, totalVotesCount);
  const isMobile = useMobile();
  const sliderRef = useRef();
  const sliderName = 'detailed_results';
  const hasVotes = votes.length > 0;
  const sliderParams: TypeSliderParams = {
    slidesToShow: 1.15,
  };

  useSlider(sliderRef, sliderParams, isMobile && hasVotes);

  return (
    <>
      <VoteProgress
        key={`vote_progress_${proposalId}`}
        votes={votes}
        proposalId={proposalId}
      />
      <GliderStylesheet />
      <ScreenReaderItemStyle>
        {i18n.t('results.static_repartition')}
      </ScreenReaderItemStyle>
      {isMobile ? (
        <div className={`${sliderName} glider-contain`}>
          <div className={`${sliderName} glider`} ref={sliderRef}>
            <DetailledItemListStyle className={`${sliderName} glider-track`}>
              {votes.map(vote => (
                <DetailledResultItem
                  key={`detail_result_${proposalId}_${vote.voteKey}`}
                  className={sliderName}
                  vote={vote}
                  votePercent={votesPercent[vote.voteKey]}
                />
              ))}
            </DetailledItemListStyle>
          </div>
        </div>
      ) : (
        <DetailledItemListStyle>
          {votes.map(vote => (
            <DetailledResultItem
              key={`detail_result_${proposalId}_${vote.voteKey}`}
              vote={vote}
              votePercent={votesPercent[vote.voteKey]}
            />
          ))}
        </DetailledItemListStyle>
      )}
    </>
  );
};
