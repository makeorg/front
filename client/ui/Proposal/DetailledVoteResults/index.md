##### DetailledVoteResults Example :
```jsx
import { proposalTypeFixture } from '../../../../shared/types/__fixtures__/proposal.fixture.js';
import { DetailledVoteResults } from './index';
  <DetailledVoteResults
    votes={proposalTypeFixture.votes}
    proposalId={proposalTypeFixture.id}
  />
```