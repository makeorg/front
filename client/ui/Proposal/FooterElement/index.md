##### ProposalFooterWithTagElement Example :
```jsx
import { proposalTypeFixture } from '../../../../shared/types/__fixtures__/proposal.fixture.js';
import { ProposalFooterWithTagElement } from './index';
  <ProposalFooterWithTagElement
    tags={proposalTypeFixture.tags}
  />
```

##### ProposalFooterWithQuestionElement Example :
```jsx
import { proposalTypeFixture } from '../../../../shared/types/__fixtures__/proposal.fixture.js';
import { ProposalFooterWithQuestionElement } from './index';
  <ProposalFooterWithQuestionElement
    question={proposalTypeFixture.question}
    consultationLink='#'
    isProposalAccepted
  />
```