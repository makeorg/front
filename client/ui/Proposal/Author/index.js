// @flow
import React from 'react';
import { i18n } from 'Shared/i18n';
import { type ProposalType } from 'Shared/types/proposal';
import {
  getOrganisationProfileLink,
  getPersonalityProfileLink,
} from 'Shared/helpers/url';
import { Avatar } from 'Client/ui/Avatar';
import { type StateRoot } from 'Shared/store/types';
import { RedLinkRouterStyle } from 'Client/ui/Elements/LinkElements';
import { ScreenReaderItemStyle } from 'Client/ui/Elements/AccessibilityElements';
import { trackClickPublicProfile } from 'Shared/services/Tracking';
import {
  TYPE_ORGANISATION,
  TYPE_PERSONALITY,
  TYPE_USER,
} from 'Shared/constants/user';
import {
  formatAuthorName,
  formatOrganisationName,
} from 'Shared/helpers/stringFormatter';
import { CertifiedIconStyle } from 'Client/ui/Proposal/DeprecatedAuthor/Styled';
import { useParams } from 'react-router-dom/cjs/react-router-dom.min';
import { useSelector } from 'react-redux';
import { matchMobileDevice } from 'Shared/helpers/styled';
import { AuthorInfosStyle, InfosWrapperStyle } from './style';

type Props = {
  /** Object with author's properties */
  proposal: ProposalType,
  /** Enable sequence context & specials styles */
  isSequence?: boolean,
};

export const ProposalAuthorAge = ({ age }) => {
  if (!age) {
    return null;
  }

  return <>{`, ${i18n.t('proposal_card.author.age', { age })}`}</>;
};

export const ProposalAuthorInformations = ({ proposal, isSequence }: Props) => {
  const { country } = useParams();
  const { author } = proposal;

  const isOrganisation = author.userType === TYPE_ORGANISATION;
  const isPersonality = author.userType === TYPE_PERSONALITY;
  const isBasicUser = author.userType === TYPE_USER;

  return (
    <>
      <ScreenReaderItemStyle>
        {i18n.t('proposal_card.author.from')}
      </ScreenReaderItemStyle>
      <InfosWrapperStyle className={isSequence && 'sequence'}>
        {isOrganisation && (
          <>
            <RedLinkRouterStyle
              onClick={() => trackClickPublicProfile(TYPE_ORGANISATION)}
              to={getOrganisationProfileLink(country, author.organisationSlug)}
            >
              {formatOrganisationName(author.organisationName)}
            </RedLinkRouterStyle>
            <CertifiedIconStyle aria-hidden focusable="false" />
          </>
        )}
        {isPersonality && (
          <>
            <RedLinkRouterStyle
              onClick={() => trackClickPublicProfile(TYPE_PERSONALITY)}
              to={getPersonalityProfileLink(country, proposal.userId)}
            >
              {formatAuthorName(author.firstName)}
            </RedLinkRouterStyle>
            <CertifiedIconStyle aria-hidden focusable="false" />
          </>
        )}
        {isBasicUser && formatAuthorName(author.firstName)}
        <ProposalAuthorAge age={author.age} />
      </InfosWrapperStyle>
    </>
  );
};

const setAvatarSize = (isMobile: boolean, isSequence: boolean) => {
  if (!isMobile && isSequence) {
    return 50;
  }

  return 36;
};

export const ProposalAuthor = ({ proposal, isSequence = false }: Props) => {
  const { author } = proposal;
  const { device } = useSelector((state: StateRoot) => state.appConfig);
  const isMobile = matchMobileDevice(device);
  const avatarSize = setAvatarSize(isMobile, isSequence);

  return (
    <AuthorInfosStyle>
      <Avatar avatarUrl={author.avatarUrl} isSequence avatarSize={avatarSize} />
      <ProposalAuthorInformations proposal={proposal} isSequence={isSequence} />
    </AuthorInfosStyle>
  );
};
