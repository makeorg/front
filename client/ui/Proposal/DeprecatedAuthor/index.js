// @flow
import React from 'react';
import { i18n } from 'Shared/i18n';
import { type ProposalType } from 'Shared/types/proposal';
import { DateHelper } from 'Shared/helpers/date';
import { Avatar } from 'Client/ui/Avatar';
import { type StateRoot } from 'Shared/store/types';
import { ScreenReaderItemStyle } from 'Client/ui/Elements/AccessibilityElements';
import { matchMobileDevice } from 'Shared/helpers/styled';
import { useSelector } from 'react-redux';
import { DATE_CAPITALIZE_LL_FORMAT } from 'Shared/constants/date';
import {
  AuthorDescriptionStyle,
  AuthorInfosStyle,
  ProposalStatusStyle,
} from './Styled';
import { ProposalAuthorInformations } from '../Author';

type Props = {
  /** Object with author's properties */
  proposal: ProposalType,
  /** Include avatar */
  withAvatar?: boolean,
  /** Include creation date */
  withCreationDate?: boolean,
  /** Include formatted proposal status */
  formattedProposalStatus?: string,
  /** Specifc design for sequence avatar */
  isSequence?: boolean,
};

export const DeprecatedProposalAuthor = ({
  proposal,
  withAvatar = false,
  withCreationDate = false,
  formattedProposalStatus = false,
  isSequence = false,
}: Props) => {
  const { author } = proposal;
  const { device } = useSelector((state: StateRoot) => state.appConfig);
  const isMobile = matchMobileDevice(device);
  return (
    <AuthorDescriptionStyle>
      <AuthorInfosStyle as="div" isSequence={isSequence}>
        {withAvatar && (
          <>
            {isSequence && (
              <Avatar
                avatarUrl={author.avatarUrl}
                isSequence={isSequence}
                avatarSize={isMobile ? 30 : 50}
              />
            )}
            {!isSequence && <Avatar avatarUrl={author.avatarUrl} />}
          </>
        )}
        <ProposalAuthorInformations
          proposal={proposal}
          isSequence={isSequence}
        />
        {withCreationDate && !!proposal.createdAt && (
          <>
            &nbsp;&bull;&nbsp;
            <ScreenReaderItemStyle>
              {i18n.t('proposal_card.author.date')}
            </ScreenReaderItemStyle>
            <time dateTime={proposal.createdAt}>
              {DateHelper.localizedAndFormattedDate(
                proposal.createdAt,
                DATE_CAPITALIZE_LL_FORMAT
              )}
            </time>
          </>
        )}
      </AuthorInfosStyle>
      {formattedProposalStatus && (
        <ProposalStatusStyle className={`status-${formattedProposalStatus}`}>
          <ScreenReaderItemStyle>
            {i18n.t('proposal_card.status.title')}
          </ScreenReaderItemStyle>
          {i18n.t(`proposal_card.status.${formattedProposalStatus}`)}
        </ProposalStatusStyle>
      )}
    </AuthorDescriptionStyle>
  );
};
