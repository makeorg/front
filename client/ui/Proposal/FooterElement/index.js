// @flow
import React from 'react';
import { i18n } from 'Shared/i18n';
import { type Tag as TypeTag } from 'Shared/types/proposal';
import { type Question as TypeQuestion } from 'Shared/types/question';
import { Tag } from 'Client/ui/Elements/Tag';
import { ProposalSeparatorStyle } from 'Client/ui/Proposal/Styled';
import {
  ProposalFooterStyle,
  PostedInLabelStyle,
  PostedInLinkStyle,
} from './Styled';

const NUMBER_OF_TAGS_TO_DISPLAY = 4;

type ProposalFooterWithTagElementProps = {
  tags: TypeTag[],
};

export const ProposalFooterWithTagElement = ({
  tags,
}: ProposalFooterWithTagElementProps) => {
  if (!tags.length) {
    return null;
  }

  return (
    <React.Fragment>
      <ProposalSeparatorStyle />
      <ProposalFooterStyle>
        {tags.slice(0, NUMBER_OF_TAGS_TO_DISPLAY).map(tag => (
          <Tag name={tag.label} key={tag.tagId} />
        ))}
      </ProposalFooterStyle>
    </React.Fragment>
  );
};

type ProposalFooterWithQuestionElementProps = {
  question: TypeQuestion,
  consultationLink: string,
  isProposalAccepted: boolean,
};

export const ProposalFooterWithQuestionElement = ({
  question,
  consultationLink,
  isProposalAccepted,
}: ProposalFooterWithQuestionElementProps) => {
  return (
    <React.Fragment>
      <ProposalSeparatorStyle />
      <ProposalFooterStyle>
        <PostedInLabelStyle>
          {i18n.t('proposal_card.posted_label')}
        </PostedInLabelStyle>
        <PostedInLinkStyle
          {...(isProposalAccepted
            ? {
                href: consultationLink,
              }
            : { as: 'span' })}
        >
          {question.wording.question}
        </PostedInLinkStyle>
      </ProposalFooterStyle>
    </React.Fragment>
  );
};
