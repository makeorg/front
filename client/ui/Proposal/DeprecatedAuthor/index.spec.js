import React from 'react';
import renderer from 'react-test-renderer';
import snapshotDiff from 'snapshot-diff';
import { proposalTypeFixture } from 'Shared/types/__fixtures__/proposal.fixture';
import { DeprecatedProposalAuthor } from 'Client/ui/Proposal/DeprecatedAuthor';
import * as redux from 'react-redux';

jest.mock('Client/ui/Elements/AccessibilityElements', () => ({
  ScreenReaderItemStyle: 'ScreenReaderItemStyle',
}));

jest.mock('Client/ui/Proposal/Author', () => ({
  ProposalAuthorInformations: 'ProposalAuthorInformations',
}));

jest.mock('Client/ui/Avatar', () => ({
  Avatar: 'Avatar',
}));

jest.mock('Client/ui/Svg/elements', () => ({
  SvgCheckedSymbol: 'SvgCheckedSymbol',
  SvgEmptyAvatar: 'SvgEmptyAvatar',
}));

jest.mock('Client/ui/Elements/LinkElements', () => ({
  RedLinkRouterStyle: 'RedLinkRouterStyle',
}));

jest.mock('Client/ui/Proposal/DeprecatedAuthor/Styled', () => ({
  AuthorDescriptionStyle: 'AuthorDescriptionStyle',
  AuthorInfosStyle: 'AuthorInfosStyle',
  AuthorSeparatorStyle: 'AuthorSeparatorStyle',
  ProposalStatusStyle: 'ProposalStatusStyle',
  CertifiedIconStyle: 'CertifiedIconStyle',
  InfosWrapperStyle: 'InfosWrapperStyle',
}));

jest.spyOn(redux, 'useSelector').mockReturnValue({ country: 'FR' });

describe('DeprecatedProposalAuthor', () => {
  it('must match the diff snapshot between default vs custom avatar', () => {
    const proposal = proposalTypeFixture;

    const authorWithoutAvatar = renderer
      .create(<DeprecatedProposalAuthor proposal={proposal} />)
      .toJSON();

    const authorWithAvatar = renderer
      .create(<DeprecatedProposalAuthor proposal={proposal} withAvatar />)
      .toJSON();

    expect(
      snapshotDiff(authorWithoutAvatar, authorWithAvatar)
    ).toMatchSnapshot();
  });
});
