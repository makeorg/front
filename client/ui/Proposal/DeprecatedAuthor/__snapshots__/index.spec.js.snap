// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeprecatedProposalAuthor must match the diff snapshot between default vs custom avatar 1`] = `
Snapshot Diff:
- First value
+ Second value

@@ -1,10 +1,11 @@
  <AuthorDescriptionStyle>
    <AuthorInfosStyle
      as="div"
      isSequence={false}
    >
+     <Avatar />
      <ProposalAuthorInformations
        author={
          Object {
            "age": 31,
            "firstName": "Mickael",
`;
